store/copr/mpp_probe.go

```diff
@@ -0,0 +1,268 @@
+	if t.recoveryTime.IsZero() {
+		t.recoveryTime = time.Now()
+	}
+}
+
+func (t *MPPSotreState) isRecovery(ctx context.Context, recoveryTTL time.Duration) bool {
+	if !t.lock.TryLock() {
+		return false
+	}
+	defer t.lock.Unlock()
+
+	t.lastLookupTime = time.Now()
+	if !t.recoveryTime.IsZero() && time.Since(t.recoveryTime) > recoveryTTL {
+		return true
+	}
+	logutil.Logger(ctx).Debug("Cannot detect store's availability "+
+		"because the current time has not recovery or wait mppStoreFailTTL",
+		zap.String("store address", t.address),
+		zap.Time("recovery time", t.recoveryTime),
+		zap.Duration("MPPStoreFailTTL", recoveryTTL))
+	return false
+}
+
+func (t MPPFailedStoreProbe) scan(ctx context.Context) {
+	defer func() {
+		if r := recover(); r != nil {
+			logutil.Logger(ctx).Warn("mpp failed store probe scan error,will restart", zap.Any("recover", r), zap.Stack("stack"))
+		}
+	}()
+
+	do := func(k, v any) {
+		address := fmt.Sprint(k)
+		state, ok := v.(*MPPSotreState)
+		if !ok {
+			logutil.BgLogger().Warn("MPPSotreState struct assert failed,will be clean",
+				zap.String("address", address))
+			t.Delete(address)
+			return
+		}
+
+		if !state.lock.TryLock() {
+			return
+		}
+		defer state.lock.Unlock()
+
+		state.detect(ctx, t.detectPeriod, t.detectTimeoutLimit)
+
+		// clean restored store
+		if !state.recoveryTime.IsZero() && time.Since(state.recoveryTime) > t.maxRecoveryTimeLimit {
+			t.Delete(address)
+			// clean store that may be obsolete
+		} else if state.recoveryTime.IsZero() && time.Since(state.lastLookupTime) > t.maxObsoletTimeLimit {
+			t.Delete(address)
+		}
+	}
+
+	f := func(k, v any) bool {
+		go do(k, v)
+		return true
+	}
+
+	metrics.TiFlashFailedMPPStoreState.WithLabelValues("probe").Set(-1) //probe heartbeat
+	t.failedMPPStores.Range(f)
+}
+
+// Add add a store when sync probe failed
+func (t *MPPFailedStoreProbe) Add(ctx context.Context, address string, tikvClient tikv.Client) {
+	state := MPPSotreState{
+		address:        address,
+		tikvClient:     tikvClient,
+		lastLookupTime: time.Now(),
+	}
+	logutil.Logger(ctx).Debug("add mpp store to failed list", zap.String("address", address))
+	t.failedMPPStores.Store(address, &state)
+}
+
+// IsRecovery check whether the store is recovery
+func (t *MPPFailedStoreProbe) IsRecovery(ctx context.Context, address string, recoveryTTL time.Duration) bool {
+	logutil.Logger(ctx).Debug("check failed store recovery",
+		zap.String("address", address), zap.Duration("ttl", recoveryTTL))
+	v, ok := t.failedMPPStores.Load(address)
+	if !ok {
+		// store not in failed map
+		return true
+	}
+
+	state, ok := v.(*MPPSotreState)
+	if !ok {
+		logutil.BgLogger().Warn("MPPSotreState struct assert failed,will be clean",
+			zap.String("address", address))
+		t.Delete(address)
+		return false
+	}
+
+	return state.isRecovery(ctx, recoveryTTL)
+}
+
+// Run a loop of scan
+// there can be only one background task
+func (t *MPPFailedStoreProbe) Run() {
+	if !t.lock.TryLock() {
```

Comment:
`lock` is used to make sure only one goroutine is triggered?

Reply:
yes

Reply:
Why not just use an atomic flag? Or maybe just reuse `isStop` is enough?

Reply:
stop are async operator,there will be risks。

Reply:
I mean use `t.isStop.CompareAndSwap(true, false)` to make sure only one call to `Run` will start to probe goroutine